src/librustc_codegen_ssa/mir/block.rs

```diff
@@ -651,6 +653,63 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         }
 
         if intrinsic.is_some() && intrinsic != Some("drop_in_place") {
+            let intrinsic = intrinsic.unwrap();
+            // Handle coverage-related terminators, if -Zinstrument-coverage is enabled.
+            // Some of these are metadata placeholder terminators with no backend code.
+            match intrinsic {
+                "coverage_counter_add" | "coverage_counter_subtract" => {
+                    let op = if intrinsic == "coverage_counter_add" {
+                        CounterOp::Add
+                    } else {
+                        CounterOp::Subtract
+                    };
+                    use coverage::coverage_counter_expression_args::*;
+                    let index = op_to_u32(&args[COUNTER_EXPRESSION_INDEX]);
+                    let lhs = op_to_u32(&args[LEFT_INDEX]);
+                    let rhs = op_to_u32(&args[RIGHT_INDEX]);
+                    let start_byte_pos = op_to_u32(&args[START_BYTE_POS]);
+                    let end_byte_pos = op_to_u32(&args[END_BYTE_POS]);
+                    debug!(
+                        "adding counter expression to coverage map: instance={:?}, index={}, {} {:?} {}, byte range {}..{}",
+                        self.instance, index, lhs, op, rhs, start_byte_pos, end_byte_pos,
+                    );
+                    bx.new_counter_expression_region(
+                        self.instance,
+                        index,
+                        lhs,
+                        op,
+                        rhs,
+                        start_byte_pos,
+                        end_byte_pos,
+                    );
+                    return; // Does not inject backend code
+                }
+                "coverage_unreachable" => {
+                    use coverage::coverage_unreachable_args::*;
+                    let start_byte_pos = op_to_u32(&args[START_BYTE_POS]);
+                    let end_byte_pos = op_to_u32(&args[END_BYTE_POS]);
+                    debug!(
+                        "adding unreachable code to coverage map: instance={:?}, byte range {}..{}",
+                        self.instance, start_byte_pos, end_byte_pos,
+                    );
+                    bx.new_unreachable_region(self.instance, start_byte_pos, end_byte_pos);
+                    return; // Does not inject backend code
```

Comment:
Note I added the additional terminators I will need to inject, to create the complete Coverage Map for LLVM, but only the actual counters (injected with `llvm.instrprof.increment` in response to the `count_code_region` call terminator) generates backend LLVM IR.

But, I don't have code in the `InstrumentCoverage` pass, yet, that would generate these new terminators, so there are no additional MIR tests yet either.